sycl/source/device.cpp

```diff
@@ -134,13 +137,77 @@ bool device::has_extension(const std::string &extension_name) const {
   return impl->has_extension(extension_name);
 }
 
+#ifdef __INTEL_PREVIEW_BREAKING_CHANGES
+template <typename Param>
+typename detail::is_device_info_desc<Param>::return_type
+device::get_info_internal() const {
+  return impl->template get_info<Param>();
+}
+
+detail::string device::get_device_info(detail::string_view Type) const {
+  std::string Info;
+  if (Type == typeid(info::device::name).name()) {
+    Info = impl->template get_info<info::device::name>();
+  } else if (Type == typeid(info::device::vendor).name()) {
+    Info = impl->template get_info<info::device::vendor>();
+  } else if (Type == typeid(info::device::driver_version).name()) {
+    Info = impl->template get_info<info::device::driver_version>();
+  } else if (Type == typeid(info::device::version).name()) {
+    Info = impl->template get_info<info::device::version>();
+  } else if (Type == typeid(info::device::profile).name()) {
+    Info = impl->template get_info<info::device::profile>();
+  } else if (Type == typeid(info::device::opencl_c_version).name()) {
+    Info = impl->template get_info<info::device::opencl_c_version>();
+  } else if (Type == typeid(info::device::ext_intel_pci_address).name()) {
+    Info = impl->template get_info<info::device::ext_intel_pci_address>();
+  } else {
+    throw sycl::invalid_parameter_error("unsupported device info requested",
+                                        PI_ERROR_INVALID_OPERATION);
+  }
+  return detail::string(Info);
+}
+
+std::vector<detail::string>
+device::get_device_info_vector(detail::string_view Type) const {
+  std::vector<std::string> Info;
+  if (Type == typeid(info::device::extensions).name()) {
+    Info = impl->template get_info<info::device::extensions>();
+  } else {
+    throw sycl::invalid_parameter_error(
+        "unsupported device info vector requested", PI_ERROR_INVALID_OPERATION);
+  }
+
+  std::vector<detail::string> Result;
+  for (std::string &Str : Info) {
+    Result.push_back(detail::string(Str.c_str()));
+  }
+  return Result;
+}
+#else
 template <typename Param>
 typename detail::is_device_info_desc<Param>::return_type
 device::get_info() const {
   return impl->template get_info<Param>();
 }
+#endif
 
 // Explicit override. Not fulfilled by #include device_traits.def below.
+#ifdef __INTEL_PREVIEW_BREAKING_CHANGES
```

Comment:
Please try to restructure this customization to avoid duplications. Something like this:
```
template <>
#ifdef ...
signature
#else 
signature2
#endif 
{
}
```

Can the same be applied elsewhere?